---
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Analysis of African Cichlid Fishes Data Adequacy

------------------------------------------------------------------------

## Introduction

The data set is taken from a [paper](https://www.nature.com/articles/s41559-020-01354-3) studying gene expression data in six different organs for 73 species of African Cichlid found in Lake Tanganyika. Specifically, the paper analyzes the expression patterns of both protein-coding genes and lncRNA to study the evolutionary dynamics associated with the rapid adaptive radiation known from these fish.They find that the rate of gene expression evolution varies between organs, and that the noncoding transcripts (lncRNA) evolve faster than coding transcripts. Interestingly, they also found that the rate of evolution accelerated later rather than earlier. Lastly, they used model fitting between EB, OU, and BM data to suggest that most of this evolution was dominated by stabilizing selection, with OU acting as the proxy for stabilizing selection. I will be analyzing this data set to see if that last point stands by testing the adequacy of each of those models for this data set.

## Summary Analysis

```{r include=FALSE}
library(tidyverse)
pval_files <- list.files("arbutus/pvals/brain_prot/")

readfile <- function(filename, fold="pvals"){
  res <- readRDS(paste0("arbutus/", fold, "/brain_prot/", filename))
  res
}

pvals_brain <- map_df(pval_files, readfile)

p_prot <- pvals_brain %>% select(!m.sig) %>% transmute(c.less = c.var <= 0.05, sv.less = s.var <= 0.05, sa.less = s.asr <= 0.05, sh.less = s.hgt <= 0.05 & !is.na(s.hgt), d.less = d.cdf <= 0.05) %>% transmute(inade = c.less + sv.less + sa.less + sh.less + d.less) %>% count(inade) %>% mutate(prop = n/sum(n)) %>% mutate(inade = as.character(inade), data = "protein")
p_lnc <- readRDS("arbutus/pvals/pvals_br_lnc") %>% select(!m.sig) %>% transmute(c.less = c.var <= 0.05, sv.less = s.var <= 0.05, sa.less = s.asr <= 0.05, sh.less = s.hgt <= 0.05 & !is.na(s.hgt), d.less = d.cdf <= 0.05) %>% transmute(inade = c.less + sv.less + sa.less + sh.less + d.less) %>% count(inade) %>% mutate(prop = n/sum(n)) %>% mutate(inade = as.character(inade), data = "lnc")

p_df <- full_join(p_prot, p_lnc) 

figure1 <- p_df %>% ggplot(aes(x = inade, y = n, fill = inade)) + geom_bar(stat = "identity") + geom_text(aes(label = round(prop, digits = 2))) +
  xlab("Number of inadequacies") + ylab("Number of genes") + ggtitle("Amount of genes by number of inadequacies in protein-coding and non-protein-coding genes") + facet_wrap(~data, scales = "free") + theme_bw() 

fit_files <- list.files("arbutus/Fits/brain_prot/")

fits_brain <- map(fit_files, function(x)readfile(filename = x, fold = "Fits"))

model_count <- function (fit) {
  ou = 0
  bm = 0
  eb = 0
  for(f in fit){
    vec <- f
    ifelse(vec$model == "OU", ou <- ou + 1, ifelse(vec$model == "BM", bm <- bm + 1, eb <- eb + 1))
  }
  df <- data.frame(OU = ou, BM = bm, EB = eb)
  b <- df %>% pivot_longer(c(OU, BM, EB), names_to = "model")
  b
}

counts_brain <- map_df(fits_brain, model_count) %>% group_by(model) %>% summarise(model = model, value = sum(value)) %>% unique()

figure2a <- counts_brain %>% ggplot(aes(model, value)) + geom_col() + labs(tag = "A") + theme_classic()
figure2b <- pvals_brain %>% select(!m.sig) %>% pivot_longer(cols = everything(), names_to = "tstat") %>%
    ggplot(aes(value)) + geom_histogram(aes(y = ..density..)) + facet_wrap(~tstat, nrow = 1) + labs(tag = "B") + theme_bw()
```

```{r, echo=FALSE}
figure1
```

**Figure 1: Using the best-fit model (chosen by AIC) shows a very low adequacy overall.** A) Total adequacy for long-non-coding RNA. B) Total adequacy for protein-coding genes. Adequacy overall is very low, with genes being only 4% adequate for lncRNA and 24% for protein-coding genes.

Overall it seems like the best-fit models did not adequately capture the data in neither lncRNA nor protein coding genes, suggesting that other model(s) may be needed.

## Results

### Initial Arbutus Analysis

```{r, echo=FALSE}
figure2a
figure2b
```

**Figure 2. Relative fit (A) and absolute fit (B) of the protein-coding genes.** Overall, a OU model fits the data the best in a relative sense, and in an absolute sense the best-fit model is quite inadequate. C.var, d.cdf, and s.asr show meaningful inadequacy.

<img src="arbutus/AIC/AIC_br_lnc.png" width="327"/>

<img src="arbutus/figures/arbutus_br_lnc.png" width="326"/>

**Figure 3. Relative fit (left) and absolute fit (right) of the lncRNA genes.** Overall, a OU model fits the data the best in a relative sense, and in an absolute sense the best-fit model is very inadequate. C.var, d.cdf, s.asr, and s.var show meaningful inadequacy.

Similarly to what the original paper found, most genes seemed to fit an OU model the best for both protein-coding and lncRNA genes. However, in an absolute sense the best fit models had high and very high numbers of inadequacies to protein-coding and lncRNA genes respectively. Because these inadequacies were related to c.var and s.asr, I hypothesize that a muti-rate BM model will better describe the data.

However, to perform fitting using a BMS model, selective regimes need to be defined. To get around that issue I will use the [Motmot](https://github.com/PuttickMacroevolution/motmot) package to identify rate shifts. This package works by testing likelihoods of a rate shift per clade. Likelihood of a rate shift higher than a certain cutoff value will be considered shifts in evolutionary rate for that gene. The tree will then be split by evolutionary regimes before and after the rate shift; taking the guesswork out of the equation.

```{r, include=FALSE}
library(motmot)
motmot_results <- readRDS("motmot1.rds")
```

```{r, echo=FALSE}
plot(motmot_results)
```

**Figure 4. Motmot analysis finds one rate shift in the tree.**

Adequacy analysis of the protein data using a multiple-rate BM model produces the following figures.

```{r, include=FALSE}
fit_files <- list.files("arbutus/multirate/Fits/")
pval_files <- list.files("arbutus/multirate/pvals/")
readfile2 <- function(folder, filename){
  res <- readRDS(paste0("arbutus/multirate/", folder , filename))
  res
}

fits_brain2 <- map(fit_files, function(x)readfile2("Fits/",filename = x))
pvals_brain2 <- map(pval_files, function(x)readfile2("pvals/",filename = x))

model_count2 <- function (fit) {
  ou = 0
  bm = 0
  eb = 0
  bms = 0
  for(f in fit){
    vec <- f
    ifelse(vec$model == "OU", ou <- ou + 1, ifelse(vec$model == "BM", bm <- bm + 1, ifelse(vec$model == "EB", eb <- eb + 1, bms <- bms + 1)))
  }
  df <- data.frame(OU = ou, BM = bm, EB = eb, BMS = bms)
  b <- df %>% pivot_longer(c(OU, BM, EB, BMS), names_to = "model")
  b
}


counts_brain_bms <- map_df(fits_brain2, model_count2) %>% group_by(model) %>% summarise(model = model, value = sum(value)) %>% unique()

figure5a <- counts_brain_bms %>% ggplot(aes(model, value)) + geom_col() + labs(tag = "A") + theme_classic()

pvals_brain_bms <- map_df(pval_files, function(x)readfile2("pvals/", filename = x))

figure5b <- pvals_brain_bms %>% select(!m.sig) %>% pivot_longer(cols = everything(), names_to = "tstat") %>%
    ggplot(aes(value)) + geom_histogram(aes(y = ..density..)) + facet_wrap(~tstat, nrow = 1) + labs(tag = "B") + theme_bw()

```

```{r, echo=FALSE}
figure5a
figure5b
```

**Figure 5. Relative fit (A) and absolute fit (B) of protein-coding genes with addition of a BMS model.** Overall, a OU model fits the data the best in a relative sense, with BMS being a close second. In an absolute sense the best-fit model is very inadequate. C.var, d.cdf, s.asr, and c.var show meaningful inadequacy even while using a BMS model.

Inadequacy of the data remains high even while using a BMS model, showing other factors may be at play causing inadequacies in c.var. I then wanted to quantify the effect of the addition of BMS.

```{r, include = FALSE}
p_uni <- pvals_brain %>% transmute(c.less = c.var <= 0.05, sv.less = s.var <= 0.05, sa.less = s.asr <= 0.05, sh.less = s.hgt <= 0.05 & !is.na(s.hgt), d.less = d.cdf <= 0.05) %>% transmute(inade = c.less + sv.less + sa.less + sh.less + d.less) %>% count(inade) %>% mutate(prop = n/sum(n)) %>% mutate(inade = as.character(inade), data = "w/o BMS")
p_bms <- pvals_brain_bms %>% transmute(c.less = c.var <= 0.05, sv.less = s.var <= 0.05, sa.less = s.asr <= 0.05, sh.less = s.hgt <= 0.05 & !is.na(s.hgt), d.less = d.cdf <= 0.05) %>% transmute(inade = c.less + sv.less + sa.less + sh.less + d.less) %>% count(inade) %>% mutate(prop = n/sum(n)) %>% mutate(inade = as.character(inade), data = "w/ BMS")

p_all <- full_join(p_uni, p_bms)

figure6 <- p_all %>% ggplot(aes(x = inade, y = n, fill = inade)) + geom_bar(stat = "identity") + geom_text(aes(label = round(prop, digits = 2))) +
  xlab("Number of inadequacies") + ylab("Number of genes") + ggtitle("Amount of genes by number of inadequacies with and without BMS models") + facet_wrap(~data, scales = "free") + theme_bw() 

```

```{r, echo=FALSE}
figure6
```

**Figure 6: Addition of a multi-rate BM process does little to lower the inadequacies found in the data, with similar proportions of inadequacies before and after usage.**

Overall this shows that with this data set there must be other considerations not due to multiple rates on different branches.
